src/Shared/test/CommandLineApplicationTests.cs

```diff
@@ -530,6 +924,91 @@ public void ArgumentSeparator(string[] input, string[] expectedRemaining, string
             Assert.Equal(expectedRemaining, app.RemainingArguments.ToArray());
         }
 
+        [Theory]
+        [InlineData(new string[0], new string[0], null, false)]
+        [InlineData(new[] { "--" }, new[] { "--" }, null, false)]
+        [InlineData(new[] { "-t", "val" }, new string[0], "val", false)]
+        [InlineData(new[] { "-t", "val", "--" }, new[] { "--" }, "val", false)]
+        [InlineData(new[] { "--top", "val", "--", "a" }, new[] { "--", "a" }, "val", false)]
+        [InlineData(new[] { "-t", "val", "--", "a", "--", "b" }, new[] { "--", "a", "--", "b" }, "val", false)]
+        [InlineData(new[] { "--help", "--" }, new string[0], null, true)]
+        [InlineData(new[] { "--version", "--" }, new string[0], null, true)]
+        public void ArgumentSeparator_TreatedAsUexpected(
+            string[] input,
+            string[] expectedRemaining,
+            string topLevelValue,
+            bool isShowingInformation)
+        {
+            var app = new CommandLineApplication(throwOnUnexpectedArg: false);
+            var optHelp = app.HelpOption("--help");
+            var optVersion = app.VersionOption("--version", "1", "1.0");
+            var optTop = app.Option("-t|--top <TOP>", "arg for command", CommandOptionType.SingleValue);
+
+            app.Execute(input);
+
+            Assert.Equal(topLevelValue, optTop.Value());
+            Assert.Equal(expectedRemaining, app.RemainingArguments.ToArray());
+            Assert.Equal(isShowingInformation, app.IsShowingInformation);
+
+            // Help and Version options never get values; parsing ends when encountered.
+            Assert.False(optHelp.HasValue());
+            Assert.False(optVersion.HasValue());
+        }
+
+        [Theory]
+        [InlineData(new[] { "--", "a", "--top", "val" }, new[] { "--", "a", "--top", "val" }, null, false)]
+        [InlineData(new[] { "--", "--help" }, new[] { "--", "--help" }, null, false)]
+        [InlineData(new[] { "--", "--version" }, new[] { "--", "--version" }, null, false)]
+        [InlineData(new[] { "unexpected", "--", "--version" }, new[] { "unexpected", "--", "--version" }, null, false)]
```

Comment:
@BrennanConroy the behaviour for this case would be the same with `AllowArgumentSeparator = true` because nothing checks what's being added to `RemainingArguments` after hitting `unexpected`. There's a `dotnet-watch` test covering this in AspNetCore. But, I doubt the quirky behaviour is intentional i.e. devs setting `AllowArgumentSeparator = true` would be surprised to find `--` in `RemainingArguments` and it may cause user-visible oddities. Suggest filing an issue somewhere.

Reply:
I'm not sure there's an actionable issue here. The dotnet watch help text is very clear that **anything** after an unexpected argument is passed to the child process.